### PR TITLE
 [MRG] Implementation of 1NN reweighting and reweighting example implementation

### PR DESCRIPTION
**Pull request by BuenoRuben**
based on this paper: https://arxiv.org/pdf/2102.02291.pdf

### CODE REVIEW

File: skada/_reweight.py [hunk 1]
```diff
@@ -671,6 +673,145 @@ def KLIEP(
...
+        The estimator object fitted on the source data.
+    weight_estimator_target_ : object
+        The estimator object fitted on the target data.
+    """
+
+    def __init__(self, base_estimator=None, laplace_smoothing=False):
+        super().__init__()
+        self.laplace_smoothing = laplace_smoothing
+        self.base_estimator = base_estimator or KNeighborsClassifier(
+            n_neighbors=1)
```
[the remark this answers is not in the document]
Comment (antoinedemathelin) [reply skada/_reweight.py:703]: Something like :
`n_neighbors=1, weights='uniform', algorithm='auto', leaf_size=30, p=2, metric='minkowski', metric_params=None, n_jobs=None` that I would pass to a KNearestNeighborClassifier object

What do you think ?

File: skada/_reweight.py [hunk 2]
```diff
@@ -671,6 +673,145 @@ def KLIEP(
...
+        return self
+
+    def get_weights(self, Xs, Xt):
+        indices_source = np.arange(Xs.shape[0])
+        estimator = clone(self.base_estimator)
+        estimator.fit(Xs, indices_source)
+        predictions = estimator.predict(Xt)
+        a = (1 if self.laplace_smoothing else 0)
+        weights = np.array(
+            [a + np.count_nonzero(predictions == i) for i in indices_source])
```
Comment (antoinedemathelin) [inline skada/_reweight.py:739]: Maybe you can fasten this part of the implementation by using `np.unique`, with something like:

```pyhton
unique, counts = np.unique(predictions, return_counts=True)
weights = np.ones(Xs.shape[0]) * float(self.laplace_smoothing)
weights[unique] += counts
```

File: skada/_reweight.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BuenoRuben) [reply skada/_reweight.py:703]: yes you are right, I m going to use those parametters instead

File: skada/_reweight.py (hunk 2 above)
Comment (antoinedemathelin) [reply skada/_reweight.py:739]: Hi @BuenoRuben,
I see that you mark this review as resolved, but it was not linked with my previous (not relevant) reviews for this part of the code. I still believe that using `np.unique` here is faster...

**Comment by BuenoRuben** [comment]
@antoinedemathelin I added KMM to the reweighting methods' example, and noticed that when fitting with:
n_training, n_source = 20, 20; it's almost instantly fitted
n_training, n_source = 30, 30; it's taking some time
n_training, n_source = 50, 50; it's taking about 2 minutes, while others are taking less than 5 sec

I don't know if your method is specifically having hight complexity, so I just prefer to tell you

**Comment by antoinedemathelin** [comment]
> @antoinedemathelin I added KMM to the reweighting methods' example, and noticed that when fitting with: n_training, n_source = 20, 20; it's almost instantly fitted n_training, n_source = 30, 30; it's taking some time n_training, n_source = 50, 50; it's taking about 2 minutes, while others are taking less than 5 sec
> 
> I don't know if your method is specifically having hight complexity, so I just prefer to tell you

Hi @BuenoRuben,
Yes, that's right, the qp solver used in KMM is pretty slow when n_source increase (note that n_source = 50 corresponds to 400 samples I think). This can be fasten by using `cvxopt`, which we will add as an option solver. I also plan to implement the Frank-Wolfe algorithm for KMM, which really speed up the optim.

As a first workaround, maybe we can reduce the default number of `max_iter` to 100 instead of 1000. The algorithm will not fully converge, but it will be faster...

File: skada/_reweight.py [hunk 3]
```diff
@@ -675,6 +676,258 @@ def KLIEP(
...
+    pipeline : sklearn pipeline
+        Pipeline containing the ReweightDensity adapter and the base estimator.
+    """
+    if base_estimator is None:
+        base_estimator = LogisticRegression().set_fit_request(sample_weight=True)
+
+    return make_da_pipeline(
+        NearestNeighborDensityAdapter(
+            laplace_smoothing=laplace_smoothing,
+            weights="uniform",
```
Comment (tgnassou) [inline skada/_reweight.py:919]: You don't put the parameters of your function but hardcode the values

File: skada/_reweight.py [hunk 4]
```diff
@@ -911,10 +1164,6 @@ def KMM(
         Number of maximum iteration before stopping the optimization.
     smooth_weights : bool, default=False
         If True, the weights are "smoothed" using the kernel function.
-
```
Comment (tgnassou) [inline skada/_reweight.py:914]: Why do you remove this? maybe a mistake

File: examples/methods/plot_reweighting.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        alpha=1,
+        ax=ax,
+        eps=0.5,
+        response_method="predict",
+        plot_method="contour",
+    )
+
+    ax.set_xticks(()), ax.set_yticks(())
+    ax.set_xlim(x_min, x_max), ax.set_ylim(y_min, y_max)
+    ax.set_title("Training with rewegihted data", fontsize=12)
```
Comment (tgnassou) [inline examples/methods/plot_reweighting.py:187]: typo: rewegihted -> reweighted. And for withtout DA it is unweighted

File: examples/methods/plot_reweighting.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+#     Illustration of the Discr. reweighting method
+# ------------------------------------------
+#
+# This estimator derive a class of predictive densities by weighting the source samples
+# when trying to maximize the log-likelihood function. Such approach is effective in
+# cases of covariate shift.
+#
+# See [3] for details.
+#
+# .. [3]    Hidetoshi Shimodaira. Improving predictive inference under
```
Comment (tgnassou) [inline examples/methods/plot_reweighting.py:258]: Is it the same as [2]

File: examples/methods/plot_reweighting.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+weights = weight_estimator.adapt(X, sample_domain=sample_domain).sample_weight[idx]
+
+create_plots(clf, weights=weights, name="1NN Reweight Density")
+
+# %%
+#     Illustration of the Kernel Mean Matching method
+# ------------------------------------------
+# .. _Kernel Mean Matching
+#
+# This example illustrates the use of KMM method [5] to correct covariate-shift.
```
Comment (tgnassou) [inline examples/methods/plot_reweighting.py:342]: You cite [5] and [6] and maybe you could explain a bit more KMM

File: examples/methods/plot_reweighting.py (hunk 6 above)
Comment (BuenoRuben) [reply examples/methods/plot_reweighting.py:256]: yes, in `_reweight.py` it is the same also

File: examples/methods/plot_reweighting.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+idx = extract_source_indices(sample_domain)
+weights = weight_estimator.adapt(X, sample_domain=sample_domain).sample_weight[idx]
+
+create_plots(clf, weights=weights, name="Reweight Density")
+
+# %%
+#     Illustration of the Gaussian reweighting method
+# ------------------------------------------
+# .. _Gaussian Reweight Density
+# See [1] for details.
```
[the remark this answers is not in the document]
Comment (rflamary) [reply examples/methods/plot_reweighting.py]: yes please

File: examples/methods/plot_reweighting.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+from skada import (
+    KLIEP,
+    KMM,
+    DiscriminatorReweightDensity,
+    GaussianReweightDensity,
+    ReweightDensity,
+    source_target_split,
+)
+from skada._reweight import NearestNeighborReweightDensity
```
Comment (rflamary) [inline examples/methods/plot_reweighting.py:31]: NearestNeighborReweightDensity should be imported in `skada.__init__.py`

File: examples/methods/plot_reweighting.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+#
+# When not using domain adaptation, the classifier won't train on
+# data that is distributed as the target sample domain, it will thus
+# not be performing optimaly.
+
+# We create a dict to store scores:
+scores_dict = {}
+
+
+def create_plots(
```
Comment (rflamary) [inline examples/methods/plot_reweighting.py:123]:
```suggestion
def plot_weights_and_classifier(
``````

File: skada/__init__.py [hunk 11]
```diff
@@ -35,6 +35,8 @@
     KMM,
     ReweightDensityAdapter,
     ReweightDensity,
+    NearestNeighborDensityAdapter,
```
Comment (rflamary) [inline skada/__init__.py:38]:
```suggestion
    NearestNeighborReweightDensityAdapter,
```

File: skada/__init__.py [hunk 12]
```diff
@@ -89,6 +91,8 @@
     "KMM",
     "ReweightDensityAdapter",
     "ReweightDensity",
+    "NearestNeighborDensityAdapter",
```
Comment (rflamary) [inline skada/__init__.py:94]: same here

**Review by rflamary — COMMENTED** [review]
Hello this is very nice (especially the example comparing the different reweighting methods). But when looking at it I really believe we Need to change the Methods name and this would be the good PR to do this.

basically I think we do not need the density in the reweighting methods especially sine many of them do not estimate a density at all (except ReweightDensity).

Basically Here are the name I sugest for the methods (and Adapter by adding Adaper at the end)

```python
DiscriminatorReweight,
GaussianReweight,
DensityReweight, # the only one using a density estimator but with reweight at the end
NearestNeighborReweight
```

**Comment by rflamary** [comment]
Hello @BuenoRuben sory for this infinitre PR ;) could you also rename `KMMAdapter ` to `KMMReweightAdapter` and same fro KLIEP ?

After that it seems everything will be OK

**Comment by BuenoRuben** [comment]
should I then rename `KMM` into `KMMReweight` then?

**Pull request merged**